src/data/releases/release.service.ts

```diff
@@ -19,57 +19,377 @@
+ */
+export async function getLatestRelease(): Promise<Release | undefined> {
+  const release = await releaseModel.getLatestRelease();
+
+  if (release) {
+    return toPojo(release);
   }
-  return toPojo(await releaseModel.create(files));
+
+  return undefined;
+}
+
+export async function updateActiveReleaseFiles(files: ReleaseFilesInput): Promise<Release> {
+  const activeRelease = await getActiveReleaseOrThrow('No active release.');
+  return toPojo(await releaseModel.updateRelease(activeRelease, { files }));
 }
 
 export async function updateActiveReleaseIndices(indices: string[]): Promise<Release> {
-  const activeRelease = await getActiveRelease();
-  if (!activeRelease) {
-    throw new Error('No active release.');
-  }
+  const activeRelease = await getActiveReleaseOrThrow('No active release.');
   return toPojo(await releaseModel.updateRelease(activeRelease, { indices }));
 }
 export async function updateActiveReleaseLabel(label: string): Promise<Release> {
-  const activeRelease = await getActiveRelease();
-  if (!activeRelease) {
-    throw new Error('No active release.');
-  }
+  const activeRelease = await getActiveReleaseOrThrow('No active release.');
   return toPojo(await releaseModel.updateRelease(activeRelease, { label }));
 }
 export async function updateActiveReleaseSnapshot(snapshot: string): Promise<Release> {
-  const activeRelease = await getActiveRelease();
-  if (!activeRelease) {
-    throw new Error('No active release.');
-  }
+  const activeRelease = await getActiveReleaseOrThrow('No active release.');
   return toPojo(await releaseModel.updateRelease(activeRelease, { snapshot }));
 }
 
-export async function publishActiveRelease(): Promise<Release> {
-  const activeRelease = await getActiveRelease();
+export async function beginCalculatingActiveRelease(): Promise<ReleaseStateChangeResponse> {
+  let activeRelease = await getActiveRelease();
+  if (!activeRelease) {
+    // Create a new release if no active release is found
+    activeRelease = toPojo(await releaseModel.create());
+  }
+
   if (!activeRelease) {
-    throw new Error('No active release.');
+    logger.error(
+      'Unable to get or create an active release for calculating. Throwing error to break the process.',
+    );
+    throw new Error('Unexpected error reading or creating active release.');
+  }
+
+  const previousState = activeRelease.state;
+
+  if (isAllowedStateTransition(activeRelease, ReleaseState.CALCULATING)) {
+    logger.info(
+      `Transitioning active release from ${previousState} to ${ReleaseState.CALCULATING}. Clearing all existing release calculations and build data.`,
+    );
+    const release = toPojo(
+      await releaseModel.updateRelease(activeRelease, {
+        state: ReleaseState.CALCULATING,
+        files: null,
+        builtAt: null,
+        calculatedAt: null,
+        label: null,
+        indices: null,
+        snapshot: null,
+        error: null,
+      }),
```

Comment:
Properly reset properties of the release in case we are re-calculating a release that was previously calculated or built.